br/pkg/lightning/backend/local/engine.go

```diff
@@ -1471,5 +1478,8 @@ func (i dbSSTIngester) ingest(metas []*sstMeta) error {
 	for _, m := range metas {
 		paths = append(paths, m.path)
 	}
+	if i.e.db == nil {
+		return errorEngineClosed
```

Comment:
after this fix, is `WithRecover` still needed? panic will break the control flow, even if we can recover from it, we may still face dead locks because some code is not executed unexpected.

Reply:
I think it's a bit dangerous that we recover from the two functions. For example, `ingestSSTLoop` has a defered WaitGroup.Done and `e.setError`. If it panics before `e.setError , after recover, WaitGroup.Wait is continued and it will see no error, checkpoint maybe forwarded wrongly.

The original purpose is to write a friendly panic log, I think it's enough to add a recover to `Lightning.run` or elsewhere.

Reply:
Oh we can't catch panic from another goroutine, so I'm really doubting about what's the expected solution for https://github.com/pingcap/tidb/issues/33524

Reply:
> Oh we can't catch panic from another goroutine, so I'm really doubting about what's the expected solution for #33524 . Maybe log something and throw the panic again?

Then there is no big difference with not recovering from panic.

I read an article https://go.dev/blog/defer-panic-and-recover and read the panic usage in `encoding/json`. After that, I think we should panic directly.